From 7150ba655c0cc08fa6ded309b81265bb672f2869 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Wed, 25 Jan 2023 11:41:40 +1000
Subject: [PATCH xserver] Xi: fix potential use-after-free in
 DeepCopyPointerClasses

CVE-2023-0494, ZDI-CAN 19596

This vulnerability was discovered by:
Jan-Niklas Sohn working with Trend Micro Zero Day Initiative

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 Xi/exevents.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Index: xorg-server-1.20.4/Xi/exevents.c
===================================================================
--- xorg-server-1.20.4.orig/Xi/exevents.c	2023-02-06 14:37:16.771887530 +0100
+++ xorg-server-1.20.4/Xi/exevents.c	2023-02-06 14:37:16.767887532 +0100
@@ -575,8 +575,10 @@
             memcpy(to->button->xkb_acts, from->button->xkb_acts,
                    sizeof(XkbAction));
         }
-        else
+        else {
             free(to->button->xkb_acts);
+            to->button->xkb_acts = NULL;
+        }
 
         memcpy(to->button->labels, from->button->labels,
                from->button->numButtons * sizeof(Atom));
